
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>8. Alpha diversity &#8212; q2book</title>
    
  <link rel="stylesheet" href="../_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/sphinx-book-theme.e7340bb3dbd8dde6db86f25597f54a1b.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.d3f166471bb80abb5163.js">

    <script id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.7d483ff0a819d6edff12ce0b1ead3928.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="9. Beta diversity" href="beta-diversity.html" />
    <link rel="prev" title="7. Taxonomic annotation and analysis of sequences" href="taxonomy.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />



  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../index.html">
  
  <img src="../_static/logo.png" class="logo" alt="logo">
  
  
  <h1 class="site-logo" id="site-title">q2book</h1>
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../front-matter/preface.html">
   Preface
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Front Matter
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../front-matter/reading.html">
   1. Reading this book
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Microbiome Bioinformatics with QIIME 2
 </span>
</p>
<ul class="current nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="getting-started.html">
   1. Getting started with using QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interactive-introduction.html">
   2. An interactive introduction to QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="metadata.html">
   3. Metadata in QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="importing.html">
   4. Importing data into QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="demultiplexing.html">
   5. Demultiplexing a sequencing run
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="denoising.html">
   6. Denoising demultiplexed sequencing data
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="taxonomy.html">
   7. Taxonomic annotation and analysis of sequences
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   8. Alpha diversity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="beta-diversity.html">
   9. Beta diversity
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="feature-table-normalization.html">
   10. Normalization of Feature Tables
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Understanding the algorithms
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../algorithms/pairwise-alignment.html">
   1. Pairwise sequence alignment
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../algorithms/database-searching.html">
   2. Sequence homology searching
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../algorithms/machine-learning.html">
   3. Machine learning in bioinformatics
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Developing with QIIME 2
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../developing/getting-started.html">
   1. Getting started with developing with QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../developing/first-plugin.html">
   2. Building a first plugin
  </a>
 </li>
</ul>
<p class="caption collapsible-parent">
 <span class="caption-text">
  Back Matter
 </span>
</p>
<ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/importing-examples.html">
   1. Appendix: Examples illustrating importing data into QIIME 2
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/biological-information.html">
   2. Appendix: Biological Information
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/advanced-metadata-formatting.html">
   3. Appendix: Advanced metadata formatting
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/microbiome-methods.html">
   4. Appendix: A brief introduction to methods in microbiome science
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/q2-origin-story.html">
   5. Appendix: History of the QIIME platform
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../back-matter/glossary.html">
   6. Glossary
  </a>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        <a class="dropdown-buttons"
            href="../_sources/using/alpha-diversity.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download notebook file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/using/alpha-diversity.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->


            <!-- Full screen (wrap in <a> to have style consistency -->
            <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                    data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                    title="Fullscreen mode"><i
                        class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/gregcaporaso/q2book/main?urlpath=tree/book/using/alpha-diversity.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i>
            Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#metrics-of-microbiome-richness">
   8.1. Metrics of microbiome richness
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#observed-features">
     8.1.1. Observed features
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#even-sampling">
     8.1.2. Even sampling
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#a-rarefied-feature-table-at-sampling-depth-10">
       8.1.2.1. A rarefied feature table at
       <code class="docutils literal notranslate">
        <span class="pre">
         sampling_depth=10
        </span>
       </code>
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#another-rarefied-feature-table-at-sampling-depth-10">
       8.1.2.2. Another rarefied feature table at
       <code class="docutils literal notranslate">
        <span class="pre">
         sampling_depth=10
        </span>
       </code>
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#a-third-rarefied-feature-table-at-sampling-depth-10">
       8.1.2.3. A third rarefied feature table at
       <code class="docutils literal notranslate">
        <span class="pre">
         sampling_depth=10
        </span>
       </code>
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#a-rarefied-feature-table-at-a-higher-sampling-depth-sampling-depth-13">
       8.1.2.4. A rarefied feature table at a higher sampling depth (
       <code class="docutils literal notranslate">
        <span class="pre">
         sampling_depth=13
        </span>
       </code>
       )
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#a-limitation-of-feature-counting">
       8.1.2.5. A limitation of feature counting
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#phylogenetic-diversity-pd">
     8.1.3. Phylogenetic Diversity (PD)
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#faith-s-pd-computed-on-our-first-rarefied-feature-table">
       8.1.3.1. Faith’s PD computed on our first rarefied feature table
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#faith-s-pd-computed-on-our-second-rarefied-feature-table">
       8.1.3.2. Faith’s PD computed on our second rarefied feature table
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#faith-s-pd-computed-on-our-third-rarefied-feature-table">
       8.1.3.3. Faith’s PD computed on our third rarefied feature table
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#what-about-chao1">
     8.1.4. What about Chao1?
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#microbiome-evenness">
   8.2. Microbiome evenness
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#shannon-diversity">
   8.3. Shannon diversity
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#alpha-rarefaction">
   8.4. Alpha rarefaction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#list-of-works-cited">
   8.5. List of works cited
  </a>
 </li>
</ul>

        </nav>
        
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="alpha-diversity">
<span id="id1"></span><h1><span class="section-number">8. </span>Alpha diversity<a class="headerlink" href="#alpha-diversity" title="Permalink to this headline">¶</a></h1>
<p>In this chapter we’ll begin to explore metrics of microbiome diversity. We’ll start with metrics of <strong>alpha diversity</strong>, which are measures of “within-sample” diversity. The way I typically think of these is as metrics that can be computed on a single sample.</p>
<p>The first sub-category of alpha diversity metric that we’ll look at will be <em>richness</em>. Richness refers to how many different types of organisms are present in a sample. For example, if we’re interested in species richness of plants in the Sonoran Desert and the Costa Rican rainforest, we could go to each, count the number of different species of plants that we observe, and have a basic measure of species richness in each environment.</p>
<div class="admonition-jargon-type-of-organism admonition">
<p class="admonition-title">Jargon: “type of organism”</p>
<p>As defined earlier, a “type of organism” or a “type of microbe” is an arbitrary taxonomic grouping, such as genus, species, strain, or even an amplicon sequence variant.</p>
</div>
<p>The next sub-category of alpha diversity metric that we’ll discuss in this chapter will be <em>evenness</em>. Evenness tells us how consistent the distribution of species abundances are in a given environment. If, for example, the most abundant plant in the Sonoran desert was roughly as common as the least abundant plant (not the case!), we would say that the evenness of plant species was high. On the other hand, if the most abundant plant was thousands of times more common than the least common plant (probably closer to the truth), then we’d say that the evenness of plant species was low.</p>
<div class="section" id="metrics-of-microbiome-richness">
<h2><span class="section-number">8.1. </span>Metrics of microbiome richness<a class="headerlink" href="#metrics-of-microbiome-richness" title="Permalink to this headline">¶</a></h2>
<p>We’ll begin by looking at two metrics of community richness. Both of these are commonly used in practice.</p>
<div class="section" id="observed-features">
<h3><span class="section-number">8.1.1. </span>Observed features<a class="headerlink" href="#observed-features" title="Permalink to this headline">¶</a></h3>
<p>Observed features is a very simple metric that can be used to quantify microbiome diversity. With this metric, we simply count the features that are observed in a given sample. A feature is considered to have been observed in a sample if it has a frequency of greater than zero. This is a <strong>qualitative</strong> diversity metric meaning that each feature is treated as being either observed or not observed. This metric doesn’t consider how many times a feature was observed.</p>
<p>Let’s define a simple feature table for this analysis:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">sample_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;4ac2&#39;</span><span class="p">,</span> <span class="s1">&#39;e375&#39;</span><span class="p">,</span> <span class="s1">&#39;4gd8&#39;</span><span class="p">]</span>
<span class="n">feature_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;B1&#39;</span><span class="p">,</span><span class="s1">&#39;B2&#39;</span><span class="p">,</span><span class="s1">&#39;B3&#39;</span><span class="p">,</span><span class="s1">&#39;B4&#39;</span><span class="p">,</span><span class="s1">&#39;B5&#39;</span><span class="p">,</span><span class="s1">&#39;A1&#39;</span><span class="p">,</span><span class="s1">&#39;E2&#39;</span><span class="p">]</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                 <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                 <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">]])</span>

<span class="n">feature_table_1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">sample_ids</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">feature_ids</span><span class="p">)</span>
<span class="n">feature_table_1</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3963eb20_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >B5</th>        <th class="col_heading level0 col5" >A1</th>        <th class="col_heading level0 col6" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3963eb20_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >5</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >5</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >2</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row0_col6" class="data row0 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3963eb20_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >3</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >5</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >1</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >4</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >4</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row1_col6" class="data row1 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3963eb20_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >5</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >0</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >5</td>
                        <td id="T_3963eb20_9dfe_11eb_847d_0022481c3017row2_col6" class="data row2 col6" >5</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>To compute the value of observed features for each sample in our feature table, we would simply count the number of non-zero counts for each sample. All of the non-zero counts are bolded in the following view of the table.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">non_zero_blue</span><span class="p">(</span><span class="n">frequency</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">frequency</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;blue&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span>
    <span class="k">return</span> <span class="s1">&#39;color: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">color</span>

<span class="n">feature_table_1</span><span class="o">.</span><span class="n">style</span><span class="o">.</span><span class="n">applymap</span><span class="p">(</span><span class="n">non_zero_blue</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
#T_3968410c_9dfe_11eb_847d_0022481c3017row0_col0,#T_3968410c_9dfe_11eb_847d_0022481c3017row0_col1,#T_3968410c_9dfe_11eb_847d_0022481c3017row0_col2,#T_3968410c_9dfe_11eb_847d_0022481c3017row1_col0,#T_3968410c_9dfe_11eb_847d_0022481c3017row1_col1,#T_3968410c_9dfe_11eb_847d_0022481c3017row1_col2,#T_3968410c_9dfe_11eb_847d_0022481c3017row1_col3,#T_3968410c_9dfe_11eb_847d_0022481c3017row1_col4,#T_3968410c_9dfe_11eb_847d_0022481c3017row2_col0,#T_3968410c_9dfe_11eb_847d_0022481c3017row2_col5,#T_3968410c_9dfe_11eb_847d_0022481c3017row2_col6{
            color:  blue;
        }#T_3968410c_9dfe_11eb_847d_0022481c3017row0_col3,#T_3968410c_9dfe_11eb_847d_0022481c3017row0_col4,#T_3968410c_9dfe_11eb_847d_0022481c3017row0_col5,#T_3968410c_9dfe_11eb_847d_0022481c3017row0_col6,#T_3968410c_9dfe_11eb_847d_0022481c3017row1_col5,#T_3968410c_9dfe_11eb_847d_0022481c3017row1_col6,#T_3968410c_9dfe_11eb_847d_0022481c3017row2_col1,#T_3968410c_9dfe_11eb_847d_0022481c3017row2_col2,#T_3968410c_9dfe_11eb_847d_0022481c3017row2_col3,#T_3968410c_9dfe_11eb_847d_0022481c3017row2_col4{
            color:  black;
        }</style><table id="T_3968410c_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >B5</th>        <th class="col_heading level0 col5" >A1</th>        <th class="col_heading level0 col6" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3968410c_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >5</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >5</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >2</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row0_col6" class="data row0 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3968410c_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >3</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >5</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >1</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >4</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >4</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row1_col6" class="data row1 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3968410c_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >5</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >0</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >5</td>
                        <td id="T_3968410c_9dfe_11eb_847d_0022481c3017row2_col6" class="data row2 col6" >5</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>Counting non-zero values for each sample would result in the following values of observed features for each sample:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">qiime2</span>
<span class="kn">import</span> <span class="nn">qiime2.plugins.diversity</span> <span class="k">as</span> <span class="nn">div</span>

<span class="n">feature_table_1a</span> <span class="o">=</span> <span class="n">qiime2</span><span class="o">.</span><span class="n">Artifact</span><span class="o">.</span><span class="n">import_data</span><span class="p">(</span><span class="s2">&quot;FeatureTable[Frequency]&quot;</span><span class="p">,</span> <span class="n">feature_table_1</span><span class="p">)</span>

<span class="n">observed_features_1a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha</span><span class="p">(</span><span class="n">feature_table_1a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;observed_features&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">observed_features_1</span> <span class="o">=</span> <span class="n">observed_features_1a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;observed-features&#39;</span><span class="p">)</span>
<span class="n">observed_features_1</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3ad71950_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >observed-features</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3ad71950_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3ad71950_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >3</td>
            </tr>
            <tr>
                        <th id="T_3ad71950_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3ad71950_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >5</td>
            </tr>
            <tr>
                        <th id="T_3ad71950_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3ad71950_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>Based on the observed features metric, we could consider samples <code class="docutils literal notranslate"><span class="pre">4ac2</span></code> and <code class="docutils literal notranslate"><span class="pre">4gd8</span></code> to have equal feature richness, and sample <code class="docutils literal notranslate"><span class="pre">e375</span></code> to have a higher feature richness. Also note that while <code class="docutils literal notranslate"><span class="pre">4ac2</span></code> and <code class="docutils literal notranslate"><span class="pre">4gd8</span></code> have the same feature richness, different features are present in the two samples. Richness tells us only about how many different features are present, but nothing about which features are present.</p>
<p>Simple counting of features, as we’re doing here, is common but there are a few limitations to be aware of.</p>
</div>
<div class="section" id="even-sampling">
<h3><span class="section-number">8.1.2. </span>Even sampling<a class="headerlink" href="#even-sampling" title="Permalink to this headline">¶</a></h3>
<p>Imagine again that we’re going on a sampling trip to count plants in the Sonoran Desert and the Costa Rican rainforest. We’re interested in getting an idea of the plant richness in each environment. In the Sonoran Desert, we survey a square kilometer area, and count 150 species of plants. In the rainforest, we survey a square meter, and count 15 species of plants. On the basis of our survey we decide that plant species richness is higher in the desert than in the rainforest. Where did we go wrong?</p>
<p>We expended a lot more sampling effort in the desert than we did in the rainforest, so it shouldn’t be surprising that we observed more species there. If we expended the same effort in the rainforest, we’d probably observe a lot more than 15 or 150 plant species, and we’d have a more sound comparison.</p>
<p>In sequencing-based studies of microbiomes, the analog of sampling area is sequencing depth. If we collect 100 sequences from one sample, and 10,000 sequences from another sample, we can’t directly compare the number of observed features across these samples because we expended a lot more sampling effort on the sample with 10,000 sequences than on the sample with 100 sequences.</p>
<p>The ideal way to normalize tables for computation of these metrics is a subject of ongoing research, and most likely differs depending on what you want to do with the feature table. We’ll cover this topic in <a class="reference internal" href="feature-table-normalization.html#feature-table-normalization"><span class="std std-ref">Normalization of Feature Tables</span></a>. At present, when computing alpha and <a class="reference internal" href="beta-diversity.html#beta-diversity"><span class="std std-ref">beta diversity</span></a> metics, the way this is typically handled is by randomly subsampling sequences without replacement at a fixed total frequency across all samples. This process is referred to as <strong>rarefaction</strong>. Continuing from the example above, if we randomly select 100 sequences to represent the sample with 10,000 sequences (i.e., we rarefy that sample to a depth of 100 sequences) we can compute its richness based on that random subsample. That richness value will serve as a more relevant comparison to the richness value for the sample that we only obtained 100 sequences from.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Rarefaction is not ideal. I think of it as a necessary evil that enables these comparisons. Our field needs to move toward better normalization techniques that are <em>accessible to users</em> to get beyond the need to rarefy data. You <strong>must</strong> rarefy or otherwise normalize your feature table data before computing alpha and beta diversity unless the metric you’re using specifically does not require this.</p>
</div>
<p>Because rarefaction involves taking a <em>random</em> subsample of sequences from each sample, rarefying the same feature table multiple times will yield different rarefied feature tables. This is sometimes managed by computing multiple rarefied feature tables, computing diversity metrics on each table, and then averaging the diversity value computed for each sample. We’ll explore this below in <a class="reference internal" href="#alpha-rarefaction"><span class="std std-ref">Alpha rarefaction</span></a>.</p>
<p>The following are three different rarefied versions of our example feature table from above, each rarefied to 10 sequences per sample. Notice that the total frequency for each sample is now ten. After rarefied, Observed Features is computed and presented for each rarefied table.</p>
<div class="margin sidebar">
<p class="sidebar-title"></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>When feature tables are rarefied in QIIME 2, features that are not observed in any samples are dropped from the table. This doesn’t impact downstream analysis and results in tables with smaller file size. You can observe this in the rarefied feature tables presented in this section.</p>
</div>
</div>
<div class="section" id="a-rarefied-feature-table-at-sampling-depth-10">
<span id="adiv-first-rarefied-table"></span><h4><span class="section-number">8.1.2.1. </span>A rarefied feature table at <code class="docutils literal notranslate"><span class="pre">sampling_depth=10</span></code><a class="headerlink" href="#a-rarefied-feature-table-at-sampling-depth-10" title="Permalink to this headline">¶</a></h4>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">qiime2.plugins.feature_table</span> <span class="k">as</span> <span class="nn">ft</span>

<span class="n">rarefied_feature_table_1a</span> <span class="o">=</span> <span class="n">ft</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">rarefy</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">feature_table_1a</span><span class="p">,</span> <span class="n">sampling_depth</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span><span class="o">.</span><span class="n">rarefied_table</span>
<span class="n">rarefied_feature_table1</span> <span class="o">=</span> <span class="n">rarefied_feature_table_1a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">rarefied_feature_table1</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >A1</th>        <th class="col_heading level0 col5" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >4</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >4</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >2</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
            </tr>
            <tr>
                        <th id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >2</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >4</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >0</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >4</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >0</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
            </tr>
            <tr>
                        <th id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >4</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >4</td>
                        <td id="T_3ae0ec96_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >2</td>
            </tr>
    </tbody></table></div></div>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_observed_features_1a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha</span><span class="p">(</span><span class="n">rarefied_feature_table_1a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;observed_features&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">rarefied_observed_features_1</span> <span class="o">=</span> <span class="n">rarefied_observed_features_1a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;observed-features&#39;</span><span class="p">)</span>
<span class="n">rarefied_observed_features_1</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3aecf270_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >observed-features</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3aecf270_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3aecf270_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >3</td>
            </tr>
            <tr>
                        <th id="T_3aecf270_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3aecf270_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >3</td>
            </tr>
            <tr>
                        <th id="T_3aecf270_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3aecf270_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
</div>
<div class="section" id="another-rarefied-feature-table-at-sampling-depth-10">
<h4><span class="section-number">8.1.2.2. </span>Another rarefied feature table at <code class="docutils literal notranslate"><span class="pre">sampling_depth=10</span></code><a class="headerlink" href="#another-rarefied-feature-table-at-sampling-depth-10" title="Permalink to this headline">¶</a></h4>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_feature_table_2a</span> <span class="o">=</span> <span class="n">ft</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">rarefy</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">feature_table_1a</span><span class="p">,</span> <span class="n">sampling_depth</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span><span class="o">.</span><span class="n">rarefied_table</span>
<span class="n">rarefied_feature_table2</span> <span class="o">=</span> <span class="n">rarefied_feature_table_2a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">rarefied_feature_table2</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >B5</th>        <th class="col_heading level0 col5" >A1</th>        <th class="col_heading level0 col6" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >4</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >4</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >2</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row0_col6" class="data row0 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >2</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >3</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >2</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >3</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row1_col6" class="data row1 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >3</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >0</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >4</td>
                        <td id="T_3af6ec3a_9dfe_11eb_847d_0022481c3017row2_col6" class="data row2 col6" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_observed_features_2a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha</span><span class="p">(</span><span class="n">rarefied_feature_table_2a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;observed_features&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">rarefied_observed_features_2</span> <span class="o">=</span> <span class="n">rarefied_observed_features_2a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;observed-features&#39;</span><span class="p">)</span>
<span class="n">rarefied_observed_features_2</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b02569c_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >observed-features</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b02569c_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b02569c_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >3</td>
            </tr>
            <tr>
                        <th id="T_3b02569c_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b02569c_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >4</td>
            </tr>
            <tr>
                        <th id="T_3b02569c_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b02569c_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
</div>
<div class="section" id="a-third-rarefied-feature-table-at-sampling-depth-10">
<h4><span class="section-number">8.1.2.3. </span>A third rarefied feature table at <code class="docutils literal notranslate"><span class="pre">sampling_depth=10</span></code><a class="headerlink" href="#a-third-rarefied-feature-table-at-sampling-depth-10" title="Permalink to this headline">¶</a></h4>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_feature_table_3a</span> <span class="o">=</span> <span class="n">ft</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">rarefy</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">feature_table_1a</span><span class="p">,</span> <span class="n">sampling_depth</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span><span class="o">.</span><span class="n">rarefied_table</span>
<span class="n">rarefied_feature_table3</span> <span class="o">=</span> <span class="n">rarefied_feature_table_3a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">rarefied_feature_table3</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >B5</th>        <th class="col_heading level0 col5" >A1</th>        <th class="col_heading level0 col6" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >4</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >5</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >1</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row0_col6" class="data row0 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >3</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >3</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >2</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >2</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row1_col6" class="data row1 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >4</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >0</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >3</td>
                        <td id="T_3b0bd0e6_9dfe_11eb_847d_0022481c3017row2_col6" class="data row2 col6" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_observed_features_3a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha</span><span class="p">(</span><span class="n">rarefied_feature_table_3a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;observed_features&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">rarefied_observed_features_3</span> <span class="o">=</span> <span class="n">rarefied_observed_features_3a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;observed-features&#39;</span><span class="p">)</span>
<span class="n">rarefied_observed_features_3</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b1733d2_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >observed-features</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b1733d2_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b1733d2_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >3</td>
            </tr>
            <tr>
                        <th id="T_3b1733d2_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b1733d2_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >4</td>
            </tr>
            <tr>
                        <th id="T_3b1733d2_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b1733d2_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
</div>
<div class="section" id="a-rarefied-feature-table-at-a-higher-sampling-depth-sampling-depth-13">
<h4><span class="section-number">8.1.2.4. </span>A rarefied feature table at a higher sampling depth (<code class="docutils literal notranslate"><span class="pre">sampling_depth=13</span></code>)<a class="headerlink" href="#a-rarefied-feature-table-at-a-higher-sampling-depth-sampling-depth-13" title="Permalink to this headline">¶</a></h4>
<p>If instead of choosing to rarefy at ten sequences per sample (which is lower than the total frequency of any of our samples) we rarefied at thirteen sequences per sample (which is higher than the total frequency of sample <code class="docutils literal notranslate"><span class="pre">4ac2</span></code>), sample <code class="docutils literal notranslate"><span class="pre">4ac2</span></code> will be dropped from the resulting feature table.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_feature_table_4a</span> <span class="o">=</span> <span class="n">ft</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">rarefy</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="n">feature_table_1a</span><span class="p">,</span> <span class="n">sampling_depth</span><span class="o">=</span><span class="mi">13</span><span class="p">)</span><span class="o">.</span><span class="n">rarefied_table</span>
<span class="n">rarefied_feature_table4</span> <span class="o">=</span> <span class="n">rarefied_feature_table_4a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="n">rarefied_feature_table4</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b2197c8_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >B5</th>        <th class="col_heading level0 col5" >A1</th>        <th class="col_heading level0 col6" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b2197c8_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >e375</th>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >3</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >2</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >1</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >4</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >3</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row0_col6" class="data row0 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3b2197c8_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >4gd8</th>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >4</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >0</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >0</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >0</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >0</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >5</td>
                        <td id="T_3b2197c8_9dfe_11eb_847d_0022481c3017row1_col6" class="data row1 col6" >4</td>
            </tr>
    </tbody></table></div></div>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_observed_features_4a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha</span><span class="p">(</span><span class="n">rarefied_feature_table_4a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;observed_features&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">rarefied_observed_features_4</span> <span class="o">=</span> <span class="n">rarefied_observed_features_4a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;observed-features&#39;</span><span class="p">)</span>
<span class="n">rarefied_observed_features_4</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b2cddf4_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >observed-features</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b2cddf4_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >e375</th>
                        <td id="T_3b2cddf4_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >5</td>
            </tr>
            <tr>
                        <th id="T_3b2cddf4_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >4gd8</th>
                        <td id="T_3b2cddf4_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
</div>
<div class="section" id="a-limitation-of-feature-counting">
<h4><span class="section-number">8.1.2.5. </span>A limitation of feature counting<a class="headerlink" href="#a-limitation-of-feature-counting" title="Permalink to this headline">¶</a></h4>
<p>Imagine that we have the same table, but some additional context about the features in the table. Specifically, imagine we’ve computed the following phylogenetic tree from our feature sequences. And, for the sake of illustration, imagine that we’ve also assigned taxonomy to each of the features and found that our samples contain representatives from the archaea, bacteria, and eukaryotes (their labels begin with <code class="docutils literal notranslate"><span class="pre">A</span></code>, <code class="docutils literal notranslate"><span class="pre">B</span></code>, and <code class="docutils literal notranslate"><span class="pre">E</span></code>, respectively, in this example).</p>
<!-- This code generated the tree image included in this section, but I haven't yet been able to get ete to install in the q2book environment.
```
import ete3
ts = ete3.TreeStyle()
ts.show_leaf_name = True
ts.scale = 250
ts.branch_vertical_margin = 15
ts.show_branch_length = True
```
--><div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">io</span> <span class="kn">import</span> <span class="n">StringIO</span>
<span class="kn">from</span> <span class="nn">skbio.tree</span> <span class="kn">import</span> <span class="n">TreeNode</span>

<span class="n">tree_1n</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="s1">&#39;((B1:0.1,B2:0.05):0.1,((B3:0.05,B4:0.1):0.1,B5:0.2):0.1,&#39;</span>
                   <span class="s1">&#39;((A1:0.1,A2:0.05):0.3,&#39;</span>
                   <span class="s1">&#39;(E1:0.1,E2:0.1):0.7):0.25);&#39;</span><span class="p">)</span>

<span class="n">tree_1</span> <span class="o">=</span> <span class="n">TreeNode</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">tree_1n</span><span class="p">)</span>
<span class="n">tree_1</span> <span class="o">=</span> <span class="n">tree_1</span><span class="o">.</span><span class="n">root_at_midpoint</span><span class="p">()</span>

<span class="n">tree_1a</span> <span class="o">=</span> <span class="n">qiime2</span><span class="o">.</span><span class="n">Artifact</span><span class="o">.</span><span class="n">import_data</span><span class="p">(</span><span class="s2">&quot;Phylogeny[Rooted]&quot;</span><span class="p">,</span> <span class="n">tree_1</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<!-- This code generated the tree image included in this section, but I haven't yet been able to get ete to install in the q2book environment.
```
t = ete3.Tree.from_skbio(tree, map_attributes=["value"])
t.render("%%inline", tree_style=ts)

```
-->
<div class="figure align-default" id="adiv-tree-1">
<img alt="../_images/adiv-tree-1.png" src="../_images/adiv-tree-1.png" />
<p class="caption"><span class="caption-number">Fig. 8.1 </span><span class="caption-text">A phylogenetic tree representing all of the features in our original feature table. (This tree isn’t intended to accurately represent the relationship between the Bacteria, Archaea, and Eukarya.)</span><a class="headerlink" href="#adiv-tree-1" title="Permalink to this image">¶</a></p>
</div>
<p>Pairing this with the table we defined above (displayed again in the cell below), and given what you now know about these features, how do you feel about the relative richness of these samples?</p>
</div>
</div>
<div class="section" id="phylogenetic-diversity-pd">
<span id="faith-pd"></span><h3><span class="section-number">8.1.3. </span>Phylogenetic Diversity (PD)<a class="headerlink" href="#phylogenetic-diversity-pd" title="Permalink to this headline">¶</a></h3>
<p>Phylogenetic Diversity (PD or Faith’s PD) is a metric of richness that was developed in the early 1990s <span id="id2">[<a class="reference internal" href="#id27"><span>Fai92</span></a>]</span>. Like many of the metrics that are used in microbiome research, it wasn’t initially designed for studying microbial communities, but rather communities of plants, animals, and other “macro-organisms” (macrobes?). Some of these metrics, including PD, do translate well to microbial community analysis, <a class="reference internal" href="#chao1"><span class="std std-ref">while others don’t</span></a>.</p>
<p>PD is computed as the sum of the branch lengths in a phylogenetic tree that are represented in a given sample. I recommend computing this by hand on the example data presented in this chapter to ensure that you understand how it works. It can help to have a piece of scratch paper and to print out a copy of the phylogenetic tree (<a class="reference internal" href="#adiv-tree-1"><span class="std std-numref">Fig. 8.1</span></a> in this example) to work through this process. It can also help to have a few colors of pencil or pen for this (one color per sample).</p>
<ol class="simple">
<li><p>For each sample in a given feature table, write down which features ids were observed in that sample on a different line. Choose a color to use to represent this sample.</p></li>
<li><p>For each feature id in your list, find that feature in the phylogenetic tree. Trace from the feature id to the root node of the tree in the current sample’s color. As you trace, write down the branch lengths of any <em>new</em> branches that you trace in the current color for this feature. (If you encounter a branch that you have traced for a different feature in this sample, don’t write it down again.)</p></li>
<li><p>When you have done this for all of the features observed in the sample, sum the lengths of the branches that you wrote down. Again, each branch length should be added only one time for each sample. The resulting sum is the Phylogenetic Diversity of the sample.</p></li>
<li><p>Repeat this for each sample in the feature table. Remember to choose a new color for each sample.</p></li>
</ol>
<p>Let’s apply this metric to the three rarefied feature tables that we computed above. For each rarefied feature table, I’ll print out what I would have written down on my scratch paper.</p>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">phylogenetic_diversity</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">table</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="c1"># Don&#39;t use this function in practice - it&#39;s untested and slow. Instead use </span>
    <span class="c1"># qiime2.plugins.diversity.actions.alpha_phylogenetic()</span>

    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Observed branch lengths for sample </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">sample_id</span><span class="p">)</span>
    <span class="n">sample_vector</span> <span class="o">=</span> <span class="n">table</span><span class="o">.</span><span class="n">T</span><span class="p">[</span><span class="n">sample_id</span><span class="p">]</span>
    <span class="n">observed_features</span> <span class="o">=</span> <span class="n">sample_vector</span><span class="o">.</span><span class="n">index</span><span class="p">[</span><span class="n">sample_vector</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()[</span><span class="mi">0</span><span class="p">]]</span>
    <span class="n">observed_nodes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="c1"># iterate over the observed features</span>
    <span class="k">for</span> <span class="n">feature_id</span> <span class="ow">in</span> <span class="n">observed_features</span><span class="p">:</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="n">feature_id</span><span class="p">)</span>
        <span class="n">observed_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">t</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">t</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">internal_node</span> <span class="ow">in</span> <span class="n">t</span><span class="o">.</span><span class="n">ancestors</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">internal_node</span><span class="o">.</span><span class="n">length</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># we&#39;ve hit the root</span>
                <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">verbose</span> <span class="ow">and</span> <span class="n">internal_node</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">observed_nodes</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="n">internal_node</span><span class="o">.</span><span class="n">length</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
                <span class="n">observed_nodes</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">internal_node</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">t</span><span class="o">.</span><span class="n">length</span> <span class="k">for</span> <span class="n">t</span> <span class="ow">in</span> <span class="n">observed_nodes</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">verbose</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">result</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="faith-s-pd-computed-on-our-first-rarefied-feature-table">
<h4><span class="section-number">8.1.3.1. </span>Faith’s PD computed on our first rarefied feature table<a class="headerlink" href="#faith-s-pd-computed-on-our-first-rarefied-feature-table" title="Permalink to this headline">¶</a></h4>
<p>The first rarefied feature table was as follows:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_feature_table1</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b34df54_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >A1</th>        <th class="col_heading level0 col5" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b34df54_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >4</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >4</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >2</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
            </tr>
            <tr>
                        <th id="T_3b34df54_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >2</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >4</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >0</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >4</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >0</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
            </tr>
            <tr>
                        <th id="T_3b34df54_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >4</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >4</td>
                        <td id="T_3b34df54_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >2</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>Working through the steps for each sample, I would have the following notes:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">rarefied_feature_table1</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">phylogenetic_diversity</span><span class="p">(</span><span class="n">tree_1</span><span class="p">,</span> <span class="n">rarefied_feature_table1</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed branch lengths for sample 4ac2
B1 0.1 0.1 0.25 0.125 
B2 0.05 
B3 0.05 0.1 0.1 

Observed branch lengths for sample e375
B1 0.1 0.1 0.25 0.125 
B2 0.05 
B4 0.1 0.1 0.1 

Observed branch lengths for sample 4gd8
B1 0.1 0.1 0.25 0.125 
A1 0.1 0.3 
E2 0.1 0.575 
</pre></div>
</div>
</div>
</div>
<p>This would result in the following vector of Phylogenetic Diversities:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd_1a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha_phylogenetic</span><span class="p">(</span><span class="n">rarefied_feature_table_1a</span><span class="p">,</span> <span class="n">tree_1a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;faith_pd&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">pd_1</span> <span class="o">=</span> <span class="n">pd_1a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Faith&#39;s PD&quot;</span><span class="p">)</span>
<span class="n">pd_1</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b438784_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >Faith's PD</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b438784_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b438784_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >0.875000</td>
            </tr>
            <tr>
                        <th id="T_3b438784_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b438784_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >0.925000</td>
            </tr>
            <tr>
                        <th id="T_3b438784_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b438784_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >1.650000</td>
            </tr>
    </tbody></table></div></div>
</div>
</div>
<div class="section" id="faith-s-pd-computed-on-our-second-rarefied-feature-table">
<h4><span class="section-number">8.1.3.2. </span>Faith’s PD computed on our second rarefied feature table<a class="headerlink" href="#faith-s-pd-computed-on-our-second-rarefied-feature-table" title="Permalink to this headline">¶</a></h4>
<p>The second rarefied feature table was as follows:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_feature_table2</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b44e46c_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >B5</th>        <th class="col_heading level0 col5" >A1</th>        <th class="col_heading level0 col6" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b44e46c_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >4</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >4</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >2</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row0_col6" class="data row0 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3b44e46c_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >2</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >3</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >2</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >3</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row1_col6" class="data row1 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3b44e46c_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >3</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >0</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >4</td>
                        <td id="T_3b44e46c_9dfe_11eb_847d_0022481c3017row2_col6" class="data row2 col6" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>Working through the steps for each sample, I would have the following notes:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">rarefied_feature_table2</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">phylogenetic_diversity</span><span class="p">(</span><span class="n">tree_1</span><span class="p">,</span> <span class="n">rarefied_feature_table2</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed branch lengths for sample 4ac2
B1 0.1 0.1 0.25 0.125 
B2 0.05 
B3 0.05 0.1 0.1 

Observed branch lengths for sample e375
B1 0.1 0.1 0.25 0.125 
B2 0.05 
B4 0.1 0.1 0.1 
B5 0.2 

Observed branch lengths for sample 4gd8
B1 0.1 0.1 0.25 0.125 
A1 0.1 0.3 
E2 0.1 0.575 
</pre></div>
</div>
</div>
</div>
<p>This would result in the following vector of Phylogenetic Diversities:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd_2a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha_phylogenetic</span><span class="p">(</span><span class="n">rarefied_feature_table_2a</span><span class="p">,</span> <span class="n">tree_1a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;faith_pd&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">pd_2</span> <span class="o">=</span> <span class="n">pd_2a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Faith&#39;s PD&quot;</span><span class="p">)</span>
<span class="n">pd_2</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b54550a_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >Faith's PD</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b54550a_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b54550a_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >0.875000</td>
            </tr>
            <tr>
                        <th id="T_3b54550a_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b54550a_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >1.125000</td>
            </tr>
            <tr>
                        <th id="T_3b54550a_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b54550a_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >1.650000</td>
            </tr>
    </tbody></table></div></div>
</div>
</div>
<div class="section" id="faith-s-pd-computed-on-our-third-rarefied-feature-table">
<h4><span class="section-number">8.1.3.3. </span>Faith’s PD computed on our third rarefied feature table<a class="headerlink" href="#faith-s-pd-computed-on-our-third-rarefied-feature-table" title="Permalink to this headline">¶</a></h4>
<p>The third rarefied feature table was as follows:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rarefied_feature_table3</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >B1</th>        <th class="col_heading level0 col1" >B2</th>        <th class="col_heading level0 col2" >B3</th>        <th class="col_heading level0 col3" >B4</th>        <th class="col_heading level0 col4" >B5</th>        <th class="col_heading level0 col5" >A1</th>        <th class="col_heading level0 col6" >E2</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >4</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row0_col1" class="data row0 col1" >5</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row0_col2" class="data row0 col2" >1</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row0_col3" class="data row0 col3" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row0_col4" class="data row0 col4" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row0_col5" class="data row0 col5" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row0_col6" class="data row0 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >3</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row1_col1" class="data row1 col1" >3</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row1_col2" class="data row1 col2" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row1_col3" class="data row1 col3" >2</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row1_col4" class="data row1 col4" >2</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row1_col5" class="data row1 col5" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row1_col6" class="data row1 col6" >0</td>
            </tr>
            <tr>
                        <th id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >4</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row2_col1" class="data row2 col1" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row2_col2" class="data row2 col2" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row2_col3" class="data row2 col3" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row2_col4" class="data row2 col4" >0</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row2_col5" class="data row2 col5" >3</td>
                        <td id="T_3b55cfe8_9dfe_11eb_847d_0022481c3017row2_col6" class="data row2 col6" >3</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>Working through the steps for each sample, I would have the following notes:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">sample_id</span> <span class="ow">in</span> <span class="n">rarefied_feature_table1</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">phylogenetic_diversity</span><span class="p">(</span><span class="n">tree_1</span><span class="p">,</span> <span class="n">rarefied_feature_table3</span><span class="p">,</span> <span class="n">sample_id</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Observed branch lengths for sample 4ac2
B1 0.1 0.1 0.25 0.125 
B2 0.05 
B3 0.05 0.1 0.1 

Observed branch lengths for sample e375
B1 0.1 0.1 0.25 0.125 
B2 0.05 
B4 0.1 0.1 0.1 
B5 0.2 

Observed branch lengths for sample 4gd8
B1 0.1 0.1 0.25 0.125 
A1 0.1 0.3 
E2 0.1 0.575 
</pre></div>
</div>
</div>
</div>
<p>This would result in the following vector of Phylogenetic Diversities:</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pd_3a</span> <span class="o">=</span> <span class="n">div</span><span class="o">.</span><span class="n">actions</span><span class="o">.</span><span class="n">alpha_phylogenetic</span><span class="p">(</span><span class="n">rarefied_feature_table_3a</span><span class="p">,</span> <span class="n">tree_1a</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;faith_pd&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">alpha_diversity</span>
<span class="n">pd_3</span> <span class="o">=</span> <span class="n">pd_3a</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;Faith&#39;s PD&quot;</span><span class="p">)</span>
<span class="n">pd_3</span><span class="o">.</span><span class="n">style</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><style  type="text/css" >
</style><table id="T_3b64f36a_9dfe_11eb_847d_0022481c3017" ><thead>    <tr>        <th class="blank level0" ></th>        <th class="col_heading level0 col0" >Faith's PD</th>    </tr></thead><tbody>
                <tr>
                        <th id="T_3b64f36a_9dfe_11eb_847d_0022481c3017level0_row0" class="row_heading level0 row0" >4ac2</th>
                        <td id="T_3b64f36a_9dfe_11eb_847d_0022481c3017row0_col0" class="data row0 col0" >0.875000</td>
            </tr>
            <tr>
                        <th id="T_3b64f36a_9dfe_11eb_847d_0022481c3017level0_row1" class="row_heading level0 row1" >e375</th>
                        <td id="T_3b64f36a_9dfe_11eb_847d_0022481c3017row1_col0" class="data row1 col0" >1.125000</td>
            </tr>
            <tr>
                        <th id="T_3b64f36a_9dfe_11eb_847d_0022481c3017level0_row2" class="row_heading level0 row2" >4gd8</th>
                        <td id="T_3b64f36a_9dfe_11eb_847d_0022481c3017row2_col0" class="data row2 col0" >1.650000</td>
            </tr>
    </tbody></table></div></div>
</div>
<p>How do these results compare to what we computed above with the Observed Features metric? It’s important to note that neither Observed Features nor Faith’s PD are <em>more correct</em> than the other. These metrics tell us different things about our samples, and depending on our interests we may want to choose one metric over the other. I often compute both of these metrics.</p>
<div class="warning admonition" id="pd-whole-tree">
<p class="admonition-title">What is PD_whole_tree?</p>
<p>You may occasionally see Faith’s PD values reported as “PD_whole_tree” in the literature. This is not the name of this metric, and it shouldn’t be used. Faith’s PD values should be reported as “Faith’s PD” or “Phylogenetic Diversity”, or “Faith’s Phylogenetic Diversity”.</p>
<p>I’ll take the blame for the “PD_whole_tree” misnomer. It came about because, in QIIME 1, we allowed the name of a function that computed PD to show up in figures that presented PD values. <code class="docutils literal notranslate"><span class="pre">PD_whole_tree</span></code> was <a class="reference external" href="https://github.com/pycogent/pycogent/blob/1.5.3-release/cogent/maths/unifrac/fast_unifrac.py#L220">an internal name used in our code</a> to indicate that we were computing Faith’s PD from a phylogenetic tree that had not been pruned to represent only the features that were observed in the feature table (i.e., it used the “whole tree”). This is a good name for the function, but it shouldn’t be presented to end-users of the software. Sorry about that!</p>
</div>
</div>
</div>
<div class="section" id="what-about-chao1">
<span id="chao1"></span><h3><span class="section-number">8.1.4. </span>What about Chao1?<a class="headerlink" href="#what-about-chao1" title="Permalink to this headline">¶</a></h3>
<p>Another metric that was widely used in microbiome research, especially in the early days, was Chao1 <span id="id3">[<a class="reference internal" href="#id29"><span>Cha84</span></a>]</span>. Chao1 tries to do something different than the other richness metrics we’ve looked at here: it attempts to project what the actual diversity of the environment being sampled is, rather than just computing the diversity of what we observed. This is very appealing, but the way that it does this is not compatible with sequencing-based approaches for studying microbiomes and it shouldn’t be used on this type of data.</p>
<p>Chao1 integrates the count of <strong>singleton features</strong> in its computation, where a singleton feature is defined as a feature that was only observed one time in a sample. The idea is that if you have observed a lot of singleton features, there are also probably a lot that you haven’t observed yet, so the actual diversity of the environment is likely higher than what you observe in the sample. If on the other hand, you observe few or no singleton features, that you have probably come closer to fully capturing the richness of the environment with your sample, so the actual diversity of the environment is close to what you observed.</p>
<p>This is a very cool approach, but with sequencing data we often don’t trust singletons as they frequently result from sequencing errors. In fact, many analysis workflows explicitly or implicitly filter singleton features out of samples. So, unfortunately, Chao1 shouldn’t be applied to microbiome sequence data because the counts of singletons are not reliably telling us anything about the environments we’re studying.</p>
<p>For an illustration of the effect of sequencing error on Faith’s PD, where it is handled well, versus its effect on the Chao1 metric, where it is handled less well, see Figure 1c and 1f of <span id="id4">[<a class="reference internal" href="#id28"><span>RK10</span></a>]</span>. (And, can you identify what’s <a class="reference internal" href="#pd-whole-tree"><span class="std std-ref">wrong</span></a> with the axis labels of Figure 1f?)</p>
</div>
</div>
<div class="section" id="microbiome-evenness">
<h2><span class="section-number">8.2. </span>Microbiome evenness<a class="headerlink" href="#microbiome-evenness" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="shannon-diversity">
<h2><span class="section-number">8.3. </span>Shannon diversity<a class="headerlink" href="#shannon-diversity" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="alpha-rarefaction">
<span id="id5"></span><h2><span class="section-number">8.4. </span>Alpha rarefaction<a class="headerlink" href="#alpha-rarefaction" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="list-of-works-cited">
<h2><span class="section-number">8.5. </span>List of works cited<a class="headerlink" href="#list-of-works-cited" title="Permalink to this headline">¶</a></h2>
<p id="id6"><dl class="citation">
<dt class="label" id="id29"><span class="brackets"><a class="fn-backref" href="#id3">Cha84</a></span></dt>
<dd><p>A Chao. Nonparametric estimation of the number of classes in a population. <em>Scand. Stat. Theory Appl.</em>, 11:265–270, 1984.</p>
</dd>
<dt class="label" id="id27"><span class="brackets"><a class="fn-backref" href="#id2">Fai92</a></span></dt>
<dd><p>Daniel P Faith. Conservation evaluation and phylogenetic diversity. <em>Biol. Conserv.</em>, 61(1):1–10, 1992.</p>
</dd>
<dt class="label" id="id28"><span class="brackets"><a class="fn-backref" href="#id4">RK10</a></span></dt>
<dd><p>Jens Reeder and Rob Knight. Rapidly denoising pyrosequencing amplicon reads by exploiting rank-abundance distributions. <em>Nat. Methods</em>, 7(9):668–669, September 2010.</p>
</dd>
</dl>
</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./using"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            
    <a class='left-prev' id="prev-link" href="taxonomy.html" title="previous page"><span class="section-number">7. </span>Taxonomic annotation and analysis of sequences</a>
    <a class='right-next' id="next-link" href="beta-diversity.html" title="next page"><span class="section-number">9. </span>Beta diversity</a>

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By J Gregory Caporaso<br/>
        
            &copy; Copyright 2020-2021.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../_static/js/index.d3f166471bb80abb5163.js"></script>


    
    <!-- Google Analytics -->
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'G-KZHV37SG3H', 'auto');
      ga('set', 'anonymizeIp', true);
      ga('send', 'pageview');
    </script>
    <script async src='https://www.google-analytics.com/analytics.js'></script>
    <!-- End Google Analytics -->
    
  </body>
</html>